packages/story-editor/src/components/form/hierarchical/index.js

```diff
@@ -142,15 +174,78 @@ const HierarchicalInput = ({
   const [inputText, setInputText] = useState('');
   const speak = useLiveRegion('assertive');
   const debouncedSpeak = useDebouncedCallback(speak, 500);
+  const checkboxListRef = useRef(null);
+
+  // Focus handling
+  const [focusedCheckboxId, setFocusedCheckboxId] = useState(-1);
+  const optionRefs = useRef({});
 
   const filteredOptionTree = useMemo(
     () => buildOptionsTree(options),
     [options]
   );
 
-  const filteredOptions = filterOptionsByLabelText(
-    filteredOptionTree,
-    inputText
+  const filteredOptions = useMemo(
+    () => filterOptionsByLabelText(filteredOptionTree, inputText),
+    [filteredOptionTree, inputText]
+  );
+
+  const flattenedOrderedOptions = useMemo(
+    () => flattenOptionsTree(filteredOptions),
+    [filteredOptions]
+  );
+
+  /**
+   * Handles listbox and checkbox focus.
+   *
+   * When listbox receives focus:
+   * - If none of the options are selected before the listbox receives focus, focus is set on the first option.
+   * - If one or more options are selected before the listbox receives focus, focus is set on the first option in the list that is selected.
+   */
+  const handleListboxFocus = useCallback(() => {
+    if (document.activeElement === checkboxListRef.current) {
+      if (!flattenedOrderedOptions.length) {
+        return;
+      }
+
+      const selectedId = flattenedOrderedOptions.find(
+        (option) => option.checked
+      )?.id;
+
+      setFocusedCheckboxId(selectedId || selectedId === 0 ? selectedId : -1);
```

Comment:
💭 wondering if using browser api reduces ref complexity here?
```suggestion
checkboxListRef.current?.querySelector('input[aria-checked="true"]')?.focus();
```

Reply:
Ended up doing something different - so this doesn't apply anymore